sql/hive-thriftserver/src/main/scala/org/apache/spark/sql/hive/thriftserver/SparkSQLCLIArguments.scala

```diff
@@ -0,0 +1,270 @@
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.sql.hive.thriftserver
+
+import scala.util.{Failure, Success, Try}
+
+import org.apache.spark.internal.Logging
+
+/**
+ * This class will check arguments used in command line.
+ * It mimics command line options from Hive 2.3
+ *
+ * @param args command line args.
+ */
+private[hive] case class SparkSQLCLIArguments(args: Array[String]) extends Logging {
+
+
+  /**
+   * No upper/lower case transformations. Exact values are required.
+   * Options without arguments.
+   */
+  val optionsMap = Map(
+    'V' -> "verbose", 'v' -> "verbose",
+    'S' -> "silent", 's' -> "silent",
+    'H' -> "help", 'h' -> "help"
+  )
+
+
+  /**
+   * No upper/lower case transformations. Exact values are required.
+   * Options with arguments.
+   */
+  val optionsArgMap = Map(
+    "--file" -> "file",
+    "-i" -> "init-file",
+    "--database" -> "database",
+    "--hiveconf" -> "hiveconf",
+    "-e" -> "quoted-query-string",
+    "-f" -> "file",
+    "--hivevar" -> "hivevar",
+    "-d" -> "define",
+    "--define" -> "define",
+    "--conf" -> "conf"
+  )
+
+  /**
+   * Usage string helper to show options for --help and  illegal arguments.
+   */
+  lazy val usage: String = {
+    var output = "Usage:\n"
+    output += optionsArgMap.map(pair => s"${pair._1} [ARG]\t${pair._2}").mkString("\n")
+    output += "\n\nOptions:\n"
+    output += optionsMap.map(pair => s"-${pair._1}|--${pair._2}\t${pair._2}").mkString("\n")
+    output
+  }
+
+  lazy val parsed: Map[String, Seq[String]] = {
+    /**
+     *
+     * @param innerList List of arguments, will be tail recursive.
+     * @param accumulator List of resulting Map of arguments found in innerList
+     * @param prevKey Previous key for next element in list to be its value.
+     *                (--conf(prevKey) "spark.conf=setting"(value))
+     * @return accumulator. (Tail-recursive)
+     */
+    @scala.annotation.tailrec
+    def argHelper(
+        innerList: List[String],
+        accumulator: Map[String, Seq[String]],
+        prevKey: Option[String]): Map[String, Seq[String]] = {
+
+      if (innerList.isEmpty) {
+        // End of args list.
+        accumulator
+      } else {
+        if (prevKey.isEmpty) {
+
+          // If current arg is not a key and also
+          // there is no previous key, should throw exception.
+          if (!innerList.head.startsWith("-")) {
```

Comment:
use .head is risky of throwing exception if innerList is empty.
suggest to use .headOption.exists(_.startsWith("-"))

Reply:
The head operation is covered with a if (innerList.isEmpty) should be safe.